### ISSUE TITLE
`import { default as ... } ...` doesn't work as of 4.8.3

### ISSUE THREAD

**Issue by alecmev** [microsoft/TypeScript#50690]
# Bug Report

### 🔎 Search Terms

`export default`, `import default as`, namespace, declaration.

### 🕗 Version & Regression Information

This changed between versions 4.8.2 and 4.8.3.

### ⏯ Playground Link

I can't provide a link because both Bug Workbench and CodeSandbox still don't have 4.8.3.

### 💻 Code

```ts
// @flatten-js/core/index.d.ts
declare namespace Flatten {
    class Polygon {
        constructor();
    }
}

export default Flatten;
```

[Complete typings](https://github.com/alexbol99/flatten-js/blob/1076c4cfe38443fb77cb4892f576824677177328/index.d.ts).

```ts
// index.ts
import { default as Flatten } from '@flatten-js/core';

const x = new Flatten.Polygon();
```

### 🙁 Actual behavior

I get the following error:

```
error TS2339: Property 'Polygon' does not exist on type ...
```

When looking at `Flatten`, it still has a `default` within, so I need to access `Polygon` with `Flatten.default.Polygon`.

### 🙂 Expected behavior

It gets unwrapped correctly, like in 4.8.2 and before. Right now I'm having to use the following workaround:

```ts
import { default as FlattenBad } from '@flatten-js/core';
const Flatten = FlattenBad as unknown as typeof FlattenBad.default;
```

**Comment by andrewbranch** [comment]
TL;DR: the library’s types are wrong.

@flatten-js/core is a CommonJS module (it ships some ESM files, but these are not exposed through any standard package.json fields). This means when you do a default import of it (or a `{ default as Whatever }` import) in Node, you get the whole `module.exports` object. @flatten-js/core’s `module.exports` is an object with roughly the same shape as `Flatten`, including a member named `default` whose value is `Flatten`.

```
❯ node --input-type=module -e 'import Flatten from "@flatten-js/core"; console.log(Flatten === Flatten.default)'
false

❯ node --input-type=module -e 'import Flatten from "@flatten-js/core"; console.log(Flatten.box === Flatten.default.box)'
true
```

The way you represent this in a `.d.ts` file is something like this:

```ts
declare function box(): any;
declare const Flatten: {
  box: typeof box;
}

export { box };
export { Flatten as default };
```

But instead, as you noted, the `.d.ts` file included says `export default Flatten`. This is not the same thing. It is, admittedly, really puzzling to try to figure out what `export default` means when you know from the file extension (`.d.ts`) and package.json `type` (omitted entirely) that you are looking at a CommonJS module, not an ES module. But we do our best and say that it represents something like `module.exports.default = Flatten`. Which _is_ something that the JS file actually has, which is great, but it’s also woefully incomplete. It’s missing every other export besides `default`, which is why, when you import it, TypeScript shows you that it has a `default` and nothing else. That’s exactly what the typings say.

Now, you may object, saying that obviously we should just resolve to the `default` member of the module record here. But that would be incorrect. Consider this perfectly correct JS/.d.ts pair:

```ts
// index.js
Object.defineProperty(exports, '__esModule', { value: true });
module.exports.foo = "foo";
module.exports.default = "default";

// index.d.ts
export declare const foo: "foo";
declare const _default: "default";
export default _default;
```

What should we say the type of `lib` is when you import it?

```ts
import lib from "my-wacky-lib";
```

By your hypothetical objection, it would look like `lib` should be `"default"`. But in fact, in Node, it’s an object containing properties `foo` and `default`. This is different from what runtimes/transpilers/bundlers who care about that `__esModule` flag would give you, so it’s quite important we get this right and warn you that Node is going to return the whole module record.

The author of @flatten-js/core was clever to make `module.exports` virtually indistinguishable from `module.exports.default`—it means that consumers will see the same API regardless of whether their module resolver cares about `__esModule`. Unfortunately, that same clever duplication was not copied into the type declaration file.

The fact that `{ default as Flatten }` worked was a bug, and was causing problems for libraries that _are_ typed correctly: https://github.com/microsoft/TypeScript/issues/49567.

**Comment by andrewbranch** [comment]
@weswigham, please double check my explanation 😅

**Comment by dkulchenko** [comment]
Chiming in here - while I understand that the old behavior working correctly may have been unintentional, this change in 4.8.3 led to a bunch of breaking changes in my code, which doesn't seem right for a point release.

The `{ default as X }` clause is useful in working around issues with types for RequireJS modules being imported into ESM apps where the dependent RequireJS module hasn't yet been updated for proper ESM support. 

For example, see the ioredis import ESM import example: https://github.com/luin/ioredis#basic-usage

This worked in 4.8.2:

```ts
import { default as Redis } from "ioredis";

const redisConn: Redis = new Redis();
```

but is completely broken in 4.8.3, needing to be rewritten as:

```ts
import { default as Redis } from "ioredis";

const redisConn: Redis.Redis = new Redis.default();
```

I had to make a bunch of similar changes to fastify plugin imports as well. 

This seems like a pretty major non-backwards-compatible change for a point release.

**Comment by alecmev** [comment]
Thank you for such a thorough explanation, @andrewbranch! I'll report this in Flatten instead.

**Comment by andrewbranch** [comment]
@dkulchenko yeah, I understand. We didn’t realize that the bug was being used as a workaround for broken typings. Knowing what I know now, I probably would have held the fix from 4.8.3, but the proverbial toothpaste is out of the tube now.

One correction, though—the typings that are broken were _always_ wrong; they didn’t just need to be updated for `nodenext`. The problem is just more observable in `nodenext`. Hopefully the problem libraries will get updated quickly.

**Comment by fatcerberus** [comment]
> the proverbial toothpaste is out of the tube now

Haven't heard this one before!  It just sounds so... prosaic, compared to more colorful expressions like the cat being out of the bag, or the genie being out of the bottle, or closing the barn door after the horse escaped...

**Comment by meteorlxy** [comment]
> needing to be rewritten as:
> 
> ```ts
> import { default as Redis } from "ioredis";
> 
> const redisConn: Redis.Redis = new Redis.default();
> ```

The workaround becomes even stranger than before 😅

**Comment by pastelmind** [comment]
I'm stuck on a similar situation. I am writing ESM code with TypeScript 4.8.3 and `moduleResolution: "Node16"`. I import a CommonJS library whose code looks like:

```js
// node_modules/some-lib/index.js
function someFunc() { /* ... */ };
module.exports = someFunc;
module.exports.default = someFunc;
```

and whose type definitions look like:

```ts
// node_modules/@types/some-lib/index.d.ts
let someFunc: ( /* ... */) => /* ... */;

export default someFunc;
```

...and when I attempt to call `someFunc()` like this, I run into type errors:

```ts
import someFunc from 'some-lib';

// error TS2349: This expression is not callable.
//  Type 'typeof import("/path/to/my/project/node_modules/@types/some-lib/index")' has no call signatures.
someFunc();
```

...even though the code runs fine in Node.js.

In this case, how can the type definition for `some-lib` be corrected without changing the JS source itself?

**Comment by andrewbranch** [comment]
It seems like they’re trying to do a UMD module

```ts
interface InitSqlJsStatic extends Function {
    (config?: SqlJsConfig): Promise<SqlJsStatic>;
    readonly default: this;
}

declare const SqlJs: InitSqlJsStatic;
declare namespace SqlJs {
    export { InitSqlJsStatic };
}

export = SqlJs;
export as namespace SqlJs;
```

**Comment by Mihai-github** [comment]
Hi, I'm having the same issue using `"typescript": "^4.8.2"` with React Native. There's any recommended suggestion on how to fix this or should I downgrade the version :(?

**Comment by andrewbranch** [comment]
@Mihai-github can you give us some more details on what you’re doing (including tsconfig settings) and what you’re seeing?

**Comment by Mihai-github** [comment]
@adamburgess Sure, thanks.
My tsconfig.json:
```
{
    "compilerOptions": {
        "allowJs": true,
        "allowSyntheticDefaultImports": true,
        "esModuleInterop": false,
        "isolatedModules": true,
        "jsx": "preserve",
        "lib": ["es6", "esnext"],
        "moduleResolution": "node",
        "resolveJsonModule": true,
        "noEmit": true,
        "target": "es6",
        "module": "esnext",
        "skipLibCheck": true,
        "forceConsistentCasingInFileNames": true,
        "baseUrl": "src"
    },
    "include": ["src/**/*", "src", "index.d.ts"],
    "exclude": [
        "node_modules",
        "babel.config.js",
        "metro.config.js",
        "jest.config.js"
    ]
}
```

My issue is the same with others with ```import {default as XYZ} from './abcd'```

When I'm trying to import the component in other files using the name given with this current version of typescript I cannot make it work because I get an error telling `Cannot read properties of undefined (reading 'XYZ')`.

**Comment by andrewbranch** [comment]
What import/re-export is giving you the problem?

**Comment by GabenGar** [comment]
How do I fix this issue for a ESM-only javascript package the types for which are provided by `DefinitelyTyped`?
The package: https://github.com/yeoman/stringify-object
The types: https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/stringify-object

**Comment by andrewbranch** [comment]
@GabenGar a package.json with `"type": "module"` needs to be added to the DefinitelyTyped package.

**Comment by nickluger** [comment]
> but is broken in 4.8.3, needing to be rewritten as:
> 
> ```ts
> import { default as Redis } from "ioredis";
> 
> const redisConn: Redis.Redis = new Redis.default();
> ```

This works for me... 🤷‍♂️
```ts
import RedisModule from "ioredis";

const Redis = RedisModule.default;
```

**Comment by Mister-Hope** [comment]
Hi there, this issue is heavily effecting my projects.

I am targeting pure esm packages with `target: es2020` (compatible with node14) and `moduleResolution: NodeNext` in my project.

There is a lot of packages in my repo with is exporting both esm files with `exports` and `module` and cjs files with `main`, but not having `type: module` or even use `type: commonjs`

I am sure that the second one is wrong, and I can only managed to fix types issues with them by adding `type: module`. My concern is:

- Is this solution correct? Does this have side effects?
- Is this internal changes being so breaking?

With just one hour, I am already sending severl prs to upsteam. including:

- https://github.com/developit/mitt/pull/174
- https://github.com/Rich-Harris/magic-string
- https://github.com/rollup/plugins/issues/1348
- https://github.com/mermaid-js/mermaid/pull/3818
- ...

The examples above are popular packages,(e.g.: mitt is getting Weekly Downloads at 1,960,326). If the whole ecosystem has a huge part that needs to change according to this, I am thinking that the change might be too breaking for a minor version.

**Comment by Mister-Hope** [comment]
I read again https://nodejs.org/dist/latest-v18.x/docs/api/packages.html#exports, it seems that my solution adding `type: module` may break cjs requires, so how shall I solve this? A `.d.mts` declaration file or what?
